src/persist-client/src/internal/gc.rs

```diff
@@ -313,164 +392,112 @@ where
-                            let _ = deleteable_batch_blobs.remove(&part.key);
-                            seqno_held_parts.insert(part.key.to_owned());
-                        }
-                    });
-                    // We only need to detect deletable rollups in the last iter
-                    // through the live_diffs loop because they accumulate in state.
-                    for (seqno, rollup) in state.collections.rollups.iter() {
-                        // SUBTLE: We only guarantee that a rollup exists for the
-                        // first live state. Anything before that is not allowed to
-                        // be used and so is free to be deleted and removed from
-                        // state.
-                        if seqno < &earliest_live_seqno {
-                            deleteable_rollup_blobs.push((*seqno, rollup.key.clone()));
-                        } else {
-                            // We iterate in order, may as well short circuit the
-                            // rollup loop.
-                            break;
-                        }
-                    }
-                    break;
+            });
+
+            for (seqno, rollup) in state.collections.rollups.iter() {
+                let previous = rollup_holds.insert(rollup.key.clone(), *seqno);
+                if previous.is_none() && *seqno <= req.new_seqno_since {
+                    // We've discovered a new rollup! It's safe to truncate up to here.
+                    report_step_timing(&machine.applier.metrics.gc.steps.apply_diff_seconds);
+                    truncate_range(
+                        &req,
+                        earliest_live_seqno,
+                        *seqno,
+                        machine,
+                        &delete_semaphore,
+                        &mut report_step_timing,
+                        &mut rollup_holds,
+                        &mut batch_part_holds,
+                    )
+                    .await;
+                    earliest_live_seqno = *seqno;
                 }
-                Ordering::Greater => {
-                    break;
+            }
+
+            if state.seqno == req.new_seqno_since {
+                // As described in the big rustdoc comment on [StateVersions], we
+                // maintain the invariant that there is always a rollup corresponding to
+                // the seqno of the first live version in consensus. So, write a new
+                // rollup at exactly req.new_seqno_since so we're free to truncate
+                // anything before it.
+                //
+                // NB: We write rollups periodically (via maintenance) to cover the case
+                // when GC is being held up by a long seqno hold (such as the 15m read
+                // lease timeouts whenever environmentd restarts).
+                report_step_timing(&machine.applier.metrics.gc.steps.apply_diff_seconds);
```

Comment:
the way report_step_timing works feels much more brittle in the new structure, but I think it's still probably in the realm of okay